symengine/expansion.cpp

```diff
@@ -0,0 +1,619 @@
+/**
+ *  \file expansion.h
+ *  Series expansions of expressions using Piranha
+ *
+ **/
+
+#include <symengine/symengine_config.h>
+
+#ifdef HAVE_SYMENGINE_PIRANHA
+
+#include <piranha/monomial.hpp>
+#include <piranha/polynomial.hpp>
+#include <piranha/mp_rational.hpp>
+#include <piranha/mp_integer.hpp>
+
+#include <stdexcept>
+#include <memory>
+
+#include <symengine/rational.h>
+#include <symengine/number.h>
+#include <symengine/symbol.h>
+#include <symengine/functions.h>
+#include <symengine/add.h>
+#include <symengine/mul.h>
+#include <symengine/pow.h>
+#include <symengine/visitor.h>
+#include <symengine/constants.h>
+
+//!TODO:
+//! - real coefficients
+//! - const
+//! - benchmark fast functions of short series (see ring_series.py)
+//! - Puiseux series (see ring_series.py)
+//! - multivariate expressions (see ring_series.py)
+//! - see "not implemented"
+//! - use piranha::symbol for polynomial x
+//! - make prec a reference and change it if necessary
+
+namespace SymEngine {
+
+using pp_t = piranha::polynomial<piranha::rational,piranha::monomial<short>>;
+
+//========== helpers ============
+static inline RCP<const Number> prat2synum(const piranha::rational& p_rat)
+{
+    mpq_class cl_rat(p_rat.get_mpq_view());
+    cl_rat.canonicalize();
+    if (cl_rat.get_den() == 1)
+        return make_rcp<const Integer>(cl_rat.get_num());
+    return make_rcp<const Rational>(cl_rat);
+}
+
+static inline piranha::rational syrat2prat(const Rational& syrat)
+{
+    piranha::integer i1(syrat.as_mpq().get_num().get_mpz_t());
+    piranha::integer i2(syrat.as_mpq().get_den().get_mpz_t());
+    piranha::rational r(i1);
+    r /= i2;
+    return r;
+}
+
+static inline piranha::rational mpqc2prat(const mpq_class& the_mpqc)
+{
+    piranha::integer i1(the_mpqc.get_num_mpz_t());
+    piranha::integer i2(the_mpqc.get_den_mpz_t());
+    piranha::rational r(i1);
+    r /= i2;
+    return r;
+}
+
+static const std::list<unsigned int>& step_list(unsigned int prec)
+{
+    static std::list<unsigned int> steps;
+    if (not steps.empty()) {
+        if (*(steps.rbegin()) == prec)
+            return steps;
+        else
+            steps.clear();
+    }
+
+    unsigned int tprec = prec;
+    while (tprec > 4) {
+        tprec = 2 + tprec/2;
+        steps.push_front(tprec);
+    }
+    steps.push_front(2);
+    steps.push_back(prec);
+    return steps;
+}
+
+pp_t series_invert(const pp_t s, const RCP<const Symbol> &var, unsigned int prec)
```

Comment:
First of all, since you are not storing `var` anywhere, this should be just:
```c++
pp_t series_invert(const pp_t s, const Symbol &var, unsigned int prec)
```
However, in all these functions, isn't it better to pass a piranha variable directly instead of a `Symbol`? Because you just call `get_name` to get a string and then you construct a Piranha variable/polynomial `var_p`. So I would suggest to change all these functions to just:
```c++
pp_t series_invert(const pp_t s, const pp_t var_p, unsigned int prec)
```

@bluescarni is it better to pass Piranha type `pp_t` by value or by reference?

If by reference, then the function signature should be changed to:
```c++
pp_t series_invert(const pp_t &s, const pp_t &var_p, unsigned int prec)
```